pkg/controller/usersignup/unapproved/cache.go

```diff
@@ -0,0 +1,74 @@
+package unapproved
+
+import (
+	"context"
+	"sort"
+	"sync"
+
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/pkg/apis/toolchain/v1alpha1"
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+var log = logf.Log.WithName("unapproved_usersignup_cache")
+
+type cache struct {
+	sync.RWMutex
+	userSignupsByCreation []string
+	client                client.Client
+}
+
+func (c *cache) getOldestPendingApproval(namespace string) *toolchainv1alpha1.UserSignup {
+	c.Lock()
+	defer c.Unlock()
+	oldest := c.getFirstExisting(namespace)
+	if oldest == nil {
+		c.loadLatest(namespace)
+		oldest = c.getFirstExisting(namespace)
+	}
+	return oldest
+}
+
+func (c *cache) loadLatest(namespace string) {
+	labels := map[string]string{toolchainv1alpha1.UserSignupStateLabelKey: toolchainv1alpha1.UserSignupStateLabelValuePending}
+	opts := client.MatchingLabels(labels)
+	userSignupList := &toolchainv1alpha1.UserSignupList{}
+	if err := c.client.List(context.TODO(), userSignupList, opts); err != nil {
```

Comment:
Well, yeah I could do that. I thought that getting the long list only once shouldn't be a big deal and can be slightly slow. That's also why I introduced the cache - to minimize such requests. 
I thought that proper sorting is worth listing all `UserSignups` when it's only once. The cache shouldn't take much memory either since it keeps only the names.

Reply:
I'm not worried about the cache. But what if we have thousands or even tens of thousands of usersignups in the list? I guess listing all of them can be pretty stressful for the API server and etcd?
Ideally we should limit usersignup creations in the reg-service so we do not overload etcd but we can take care of it later.

Reply:
I understood your point that you don't want to overload the API server and etcd. 
The list of the existing UserSignups should be theoretically stored in the cache of the client - as we saw in the @xcoulon experiments and testing - so it should not really reach the API server nor etcd :man_shrugging: 
Anyway, I don't want to speculate on something I don't fully understand :-). Give me some time to think about it if I find some other way.

btw, wouldn't be tens of thousands unapproved user signups a sign that we should really extend our offering and clusters? :smiley:

Reply:
@MatousJobanek let's merge this PR as is and think about improving the wait list control separately. Maybe a proper fix for this problem would be to add some limitations in the reg-service.